### PR TITLE
couchdb: add support for version 2.0.0

### PR DESCRIPTION
**Pull request by timor**
###### Motivation for this change
Support 2.0.0 CouchDB release

Version 2.0.0 is installed as a separate package called "couchdb2".
When setting the config option "package" attribute to pkgs.couchdb2, a
corresponding service configuration will be generated.  If a previous
1.6 installation exists, the databases can still be found on the local
port (default: 5986) and can be replicated from there.

Note that single-node or cluster setup still needs to be configured
manually, as described in
http://docs.couchdb.org/en/2.0.0/install/index.html.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by timor** [comment]
@LnL7 ping

File: nixos/modules/services/databases/couchdb.nix [hunk 1]
```diff
@@ -4,20 +4,29 @@ with lib;
 
 let
   cfg = config.services.couchdb;
-  configFile = pkgs.writeText "couchdb.ini"
+  useVersion2 = strings.hasPrefix "2" (builtins.parseDrvName cfg.package.name).version;
```
Comment (globin) [inline nixos/modules/services/databases/couchdb.nix]: There is `versionOlder` and `versionAtLeast` in lib. (https://github.com/NixOS/nixpkgs/blob/master/lib/strings.nix#L343-L363)

File: nixos/modules/services/databases/couchdb.nix [hunk 2]
```diff
@@ -151,6 +160,9 @@ in {
       description = "CouchDB Server";
       wantedBy = [ "multi-user.target" ];
 
+      # erlang crashes without sh in path
+      path = mkIf useVersion2 [ pkgs.bash ];
```
Comment (globin) [inline nixos/modules/services/databases/couchdb.nix]: Is it possible to wrap the binary to have it in path?

File: nixos/modules/services/databases/couchdb.nix [hunk 3]
```diff
@@ -170,11 +182,20 @@ in {
         fi
         '';
 
+      environment = mkIf useVersion2 {
+	# we are actually specifying 4 configuration files:
```
Comment (globin) [inline nixos/modules/services/databases/couchdb.nix]: indentation is off

File: pkgs/servers/http/couchdb/2.0.0.nix [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    patch bin/rebar <<EOF
+    1c1
+    < #!/usr/bin/env escript
+    ---
+    > #!${coreutils}/bin/env escript
+    EOF
+
+  '';
+
+  # HACK: this should probably not be hardcoded...
```
Comment (globin) [inline pkgs/servers/http/couchdb/2.0.0.nix]: Why is this needed?

File: pkgs/servers/http/couchdb/2.0.0.nix [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  version = "2.0.0";
+
+  src = fetchurl {
+    url = "mirror://apache/couchdb/source/${version}/apache-${name}.tar.gz";
+    sha256 = "1jkfx6g9anrgmkhrkcn50axcamragranwsciw1rhmi86rglkrbyc";
+  };
+
+  buildInputs = [ erlang icu openssl spidermonkey ];
+
+  phases = ["unpackPhase" "patchPhase" "configurePhase" "buildPhase" "installPhase"];
```
Comment (globin) [inline pkgs/servers/http/couchdb/2.0.0.nix]: You don't need to specify the phases. Why isn't `fixupPhase` being run?

**Review by globin — CHANGES_REQUESTED** [review]

**Comment by globin** [comment]
Some minor remarks but looks good otherwise

File: pkgs/servers/http/couchdb/2.0.0.nix [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  patches = [ ./jsapi.patch ];
+  postPatch = ''
+    patch src/couch/rebar.config.script <<EOF
+    74c74
+    <                 {"-DHAVE_CURL -I/usr/local/include", "-DHAVE_CURL -lmozjs185 -lcurl"};
+    ---
+    >                 {"-DHAVE_CURL -I/usr/local/include ''$NIX_CFLAGS_COMPILE", "-DHAVE_CURL -lmozjs185 -lcurl"};
+    EOF
+
+    patch bin/rebar <<EOF
```
Comment (LnL7) [inline pkgs/servers/http/couchdb/2.0.0.nix]: can you use `substituteInPlace` for this?

File: nixos/modules/services/databases/couchdb.nix [hunk 7]
```diff
@@ -130,7 +139,7 @@ in {
 
       configFile = mkOption {
         type = types.string;
-        default = "/var/lib/couchdb/couchdb.ini";
+        default = if useVersion2 then "/var/lib/couchdb/local.ini" else "/var/lib/couchdb/couchdb.ini";
```
Comment (LnL7) [inline nixos/modules/services/databases/couchdb.nix]: I think it's preferred to use `mkDefault` in the config section for dynamic defaults like this.

File: pkgs/top-level/all-packages.nix [hunk 8]
```diff
@@ -10287,6 +10287,12 @@ with pkgs;
     erlang = erlangR16;
   };
 
+  couchdb2 = callPackage ../servers/http/couchdb/2.0.0.nix {
+    spidermonkey = spidermonkey_185;
+    python = python27;
+    erlang = erlangR16;
```
Comment (LnL7) [inline pkgs/top-level/all-packages.nix]: This seems rather old, it only works with `R16`?

File: nixos/modules/services/databases/couchdb.nix (hunk 7 above)
Comment (timor) [inline nixos/modules/services/databases/couchdb.nix]: I wanted to use that, but in this case the default depends on another config option. Is that possible?

Comment (timor) [inline nixos/modules/services/databases/couchdb.nix]: ahhh nevermind, wrong option.

Comment (LnL7) [inline nixos/modules/services/databases/couchdb.nix]:
```

```

Comment (LnL7) [inline nixos/modules/services/databases/couchdb.nix]: I mean something like this
```
  config = {
    services.couchdb.configFile = mkDefault
      (if useVersion2 then "/var/lib/couchdb/local.ini" else "/var/lib/couchdb/couchdb.ini");
  };
```

File: pkgs/top-level/all-packages.nix (hunk 8 above)
Comment (LnL7) [inline pkgs/top-level/all-packages.nix]: The documentation mentions `Erlang OTP (>=R61B03, =<19.x)`

File: nixos/modules/services/databases/couchdb.nix (hunk 1 above)
Comment (timor) [inline nixos/modules/services/databases/couchdb.nix]: Using versionAtLeast would include any later versiony, not only 2.x builds.  Is that a problem?

File: pkgs/servers/http/couchdb/2.0.0.nix (hunk 6 above)
Comment (timor) [inline pkgs/servers/http/couchdb/2.0.0.nix]: Can I use that to only patch that exact line?

Comment (timor) [inline pkgs/servers/http/couchdb/2.0.0.nix]: used substituteInPlace for the CFLAGs in the rebar script

**Comment by timor** [comment]
@globin should I squash all the commits into one and force-push them to this PR?

**Comment by offlinehacker** [comment]
Yes, please do that.

**Comment by timor** [comment]
I think I addressed everything.

**Comment by timor** [comment]
@globin, @lnl7, @offlinehacker:

Is there anything left for me to change on this before it can be merged?

File: pkgs/servers/http/couchdb/2.0.0.nix [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    sha256 = "1jkfx6g9anrgmkhrkcn50axcamragranwsciw1rhmi86rglkrbyc";
+  };
+
+  buildInputs = [ erlang icu openssl spidermonkey makeWrapper ];
+
+  patches = [ ./jsapi.patch ];
+  postPatch = ''
+    substituteInPlace src/couch/rebar.config.script --replace '-DHAVE_CURL -I/usr/local/include' "-DHAVE_CURL -I/usr/local/include $NIX_CFLAGS_COMPILE"
+
+    patch bin/rebar <<EOF
```
Comment (LnL7) [inline pkgs/servers/http/couchdb/2.0.0.nix]: I think this also works
```
substituteInPlace bin/rebase \
    --replace '#!/usr/bin/env' '#!${coreutils}/bin/env'
```

**Review by LnL7 — COMMENTED** [review]
Is there a test for couchdb, if so could you add one for `couchdb2`?

File: pkgs/servers/http/couchdb/2.0.0.nix (hunk 9 above)
Comment (timor) [inline pkgs/servers/http/couchdb/2.0.0.nix]: It does not work.  "bin/rebar" is basically a bytecode archive  file with a shebang in front to call the interpreter.  For some reason that I could not debug, using substituteInPlace does not work in the same way as using patch, and causes the resulting file to be not executable correctly.  For that reason, I kept it that way.

**Comment by timor** [comment]
@LnL7 Sadly, I am not familiar with tests.  I did not see one for couchdb in nixos/tests, and there is no checkPhase in the original package either.

File: pkgs/servers/http/couchdb/2.0.0.nix (hunk 9 above)
Comment (LnL7) [inline pkgs/servers/http/couchdb/2.0.0.nix]: Ok. I'm pretty sure I asked that before 😄 .

**Review by LnL7 — APPROVED** [review]
